Add HTTPConnect header support to Upstream

pwoelfle opened the pull request:
# Description

Added `httpConnectHeaders` on `Upstream` to be able to define HTTP headers sent during HTTPConnect with the proxy.

# Context

In an enterprise landscape, proxy servers are secured with authentication mechanisms like basic auth and others. The authentication must happen during the HTTPConnect using the `Proxy-Authorization` HTTP header (see https://github.com/solo-io/gloo/issues/5859). 

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/5859

kdorosh commented on test/e2e/http_tunneling_test.go:
```diff
@@ -259,12 +261,57 @@ var _ = Describe("tunneling", func() {
+
+			It("should not proxy", func() {
+				jsonStr := `{"value":"Hello, world!"}`
+				testReq := testRequest(jsonStr, http.StatusServiceUnavailable)
```
why does the client see a 503 error? i'd expect the 401 to be returned

pwoelfle replied:
Unfortunately, I couldn't figured out where the HTTP 503 was coming from.

I thought it is somehow correct that the service is unavailable since the HTTP Connect which establish the proxied connection does not work - so the real target service is "unavailable" from the client's point of view.

I'm happy for any suggestion.

kdorosh replied:
I think the error will be clear if you look at the debug envoy logs.. I bet we aren't handling / propagating the 401 error properly across the internal envoy listener (the pipe)

kdorosh replied:
if you can add the debug envoy logs to a gist im happy to look at them here with you or in our public slack

pwoelfle replied:
@kdorosh Thank you for your feedback. I've just put the log of the unauthenticated test case ("with Proxy Authorization" -> "using invalid credentials") in a gist: https://gist.github.com/pwoelfle/0dd64d03f377646253effeaf6987222e

pwoelfle replied:
I took a deeper look into the Envoy TcpProxy implementation to understand what happens during the HTTP Connect.

The `DecoderShim` handles the HTTP/1 and HTTP/2 responses of the HTTP Connect request in the [`decodeHeaders`](https://github.com/envoyproxy/envoy/blob/v1.23.0/source/common/tcp_proxy/upstream.h#L168-L175).

As described in [`Http1Upstream::isValidResponse`](https://github.com/envoyproxy/envoy/blob/v1.23.0/source/common/tcp_proxy/upstream.cc#L326-L332), according to the RFC7231 only 2xx responses must be considered as successful.
In our test case here, a 4xx status code is returned, so the [`HttpUpstream::resetEncoder`](https://github.com/envoyproxy/envoy/blob/v1.23.0/source/common/tcp_proxy/upstream.cc#L116-L134) is called to reset the connection by handling it as a pool failure.

From this perspective, the connection could not establish which is interpreted as service unavailable. I don't see any way to forward the HTTP Connect response status code.

kdorosh reviewed (changes requested):
thanks for opening this pr and making a contribution! changes look very reasonable; just one concern about the UX for clients

pwoelfle commented on test/e2e/http_tunneling_test.go:
```diff
@@ -327,6 +374,14 @@ func connectProxy(w http.ResponseWriter, r *http.Request) {
+	if proxyAuth := r.Header.Get("Proxy-Authorization"); proxyAuth != "" {
+		fmt.Fprintf(GinkgoWriter, "proxy authorization: %s\n", proxyAuth)
+		if username, password := parseBasicAuth(proxyAuth); username != "test" || password != "secret" {
+			w.WriteHeader(http.StatusProxyAuthRequired)
```
Switched to [HTTP 407 Proxy Authentication Required](https://httpwg.org/specs/rfc7235.html#status.407) since it is fits better in our example.

kdorosh commented:
/test

kdorosh reviewed (dismissed):
thank you!

pwoelfle commented:
@kdorosh Thanks for approving. 

For some reason the build-bot marks the CI build as failed. Can you give me a hint?

kdorosh commented:
it looks like our version is wrong:
```
Step #12 - "docker-push-extended": GO111MODULE=on CGO_ENABLED=0 GOARCH=amd64 GOOS=linux go build -ldflags="-X github.com/solo-io/gloo/pkg/version.Version=-6946" -gcflags=all="-N -l" -o /workspace/gloo/_output/projects/discovery/discovery-linux-amd64 projects/discovery/cmd/main.go
```

see https://github.com/solo-io/gloo/blob/c2d91d79b097960e2e2842dbf3a78d1b3e2232ea/Makefile#L32-L43

Those `git describe` commands don't work from your fork, but we can short circuit and ignore them completely by adding this right before those lines (which should fix your PR):

```
# ensure we have a valid version from a forked repo, so community users can submit PRs
ORIGIN_URL := "$(shell git remote get-url origin)"
UPSTREAM_ORIGIN_URL := "git@github.com:solo-io/gloo.git"
ifneq ($(ORIGIN_URL),$(UPSTREAM_ORIGIN_URL))
	VERSION := 0.0.0-fork
	CREATE_TEST_ASSETS := "false"
endif
```

sorry about the hassle here!

pwoelfle commented:
@kdorosh Thank you for the makefile version fix. I've just pushed this change. Can you please trigger the CI once again?

pwoelfle commented:
@kdorosh I get the following error from your changelog-bot:
```
Expected version v1.13.0 to be next changelog version, found v1.12.8
```

Is this the case since it is a new feature and your policy does not allow to ship new releases on older versions? Do I have to merge it into the `master`.

kdorosh commented:
yeah, can you update it to master? we can handle the backport. once updated i can kick off ci

pwoelfle commented:
@kdorosh I've just rebased to `master` and adjust the pull-request target branch also to `master`. I've changed the changelog file to `v1.13.0` as suggested by the changelog-bot.

kdorosh commented:
bot is wrong; can we use `1.13.0-beta7` as last release is beta6 per https://github.com/solo-io/gloo/releases

pwoelfle commented:
@kdorosh Changelog file moved to `v1.13.0-beta7` 👍🏻

kdorosh commented:
/test

kdorosh reviewed (approved):
thank you for the contribution!

pwoelfle commented:
@kdorosh Thanks for your help!

kdorosh commented:
sorry, one last version change for tests on a fork:

```
diff --git a/install/test/helm_suite_test.go b/install/test/helm_suite_test.go
index 0d4a2e51..998980b8 100644
--- a/install/test/helm_suite_test.go
+++ b/install/test/helm_suite_test.go
@@ -8,6 +8,7 @@ import (
        "os"
        "os/exec"
        "path"
+       "strings"
        "testing"
        "text/template"
 
@@ -22,11 +23,9 @@ import (
        . "github.com/onsi/ginkgo"
        . "github.com/onsi/gomega"
        "github.com/solo-io/gloo/pkg/cliutil/helm"
-       glooVersion "github.com/solo-io/gloo/pkg/version"
        "github.com/solo-io/gloo/projects/gloo/cli/pkg/cmd/install"
        "github.com/solo-io/gloo/projects/gloo/pkg/defaults"
        "github.com/solo-io/go-utils/testutils"
-       "github.com/solo-io/go-utils/versionutils/git"
        . "github.com/solo-io/k8s-utils/manifesttestutils"
        "helm.sh/helm/v3/pkg/action"
        "helm.sh/helm/v3/pkg/chart/loader"
@@ -46,15 +45,7 @@ func TestHelm(t *testing.T) {
 }
 
 var _ = BeforeSuite(func() {
-       version = os.Getenv("TAGGED_VERSION")
-       if !glooVersion.IsReleaseVersion() {
-               gitInfo, err := git.GetGitRefInfo("./")
-               Expect(err).NotTo(HaveOccurred())
-               // remove the "v" prefix
-               version = gitInfo.Tag[1:]
-       } else {
-               version = version[1:]
-       }
+       version = MustGetVersion()
        pullPolicy = v1.PullIfNotPresent
        // generate the values.yaml and Chart.yaml files
        MustMake(".", "-C", "../../", "generate-helm-files", "-B")
@@ -96,6 +87,30 @@ func MustMake(dir string, args ...string) {
        ExpectWithOffset(1, err).NotTo(HaveOccurred())
 }
 
+func MustMakeReturnStdout(dir string, args ...string) string {
+       makeCmd := exec.Command("make", args...)
+       makeCmd.Dir = dir
+
+       var stdout bytes.Buffer
+       makeCmd.Stdout = &stdout
+
+       makeCmd.Stderr = GinkgoWriter
+       err := makeCmd.Run()
+
+       ExpectWithOffset(1, err).NotTo(HaveOccurred())
+
+       return stdout.String()
+}
+
+func MustGetVersion() string {
+       output := MustMakeReturnStdout(".", "-C", "../../", "print-VERSION") // use print-VERSION so version matches on forks
+       // sample output
+       // make: Entering directory '/var/home/kdorosh/git/forks/gloo'\n0.0.0-fork\nmake: Leaving directory '/var/home/kdorosh/git/forks/gloo'\n"
+       lines := strings.Split(output, "\n")
+       Expect(len(lines)).To(BeNumerically(">", 2))
+       return lines[1]
+}
+
```

fixes it for me

pwoelfle commented:
@kdorosh Patch applied.

kdorosh commented:
/test

kdorosh commented:
new diff to apply :sweat: terrible tech debt in this test so sorry about this

```
diff --git a/projects/gloo/pkg/plugins/utils/update_upstream_spec.go b/projects/gloo/pkg/plugins/utils/update_upstream_spec.go
index 1064caa1..e272ccdf 100644
--- a/projects/gloo/pkg/plugins/utils/update_upstream_spec.go
+++ b/projects/gloo/pkg/plugins/utils/update_upstream_spec.go
@@ -63,6 +63,10 @@ func UpdateUpstream(original, desired *v1.Upstream) {
                desired.HttpProxyHostname = original.GetHttpProxyHostname()
        }
 
+       if desired.GetHttpConnectHeaders() == nil {
+               desired.HttpConnectHeaders = original.GetHttpConnectHeaders()
+       }
+
        if desiredSubsetMutator, ok := desired.GetUpstreamType().(v1.SubsetSpecMutator); ok {
                if desiredSubsetMutator.GetSubsetSpec() == nil {
                        desiredSubsetMutator.SetSubsetSpec(original.GetUpstreamType().(v1.SubsetSpecGetter).GetSubsetSpec())
diff --git a/projects/gloo/pkg/plugins/utils/update_upstream_spec_test.go b/projects/gloo/pkg/plugins/utils/update_upstream_spec_test.go
index ba58d18e..ea121aa3 100644
--- a/projects/gloo/pkg/plugins/utils/update_upstream_spec_test.go
+++ b/projects/gloo/pkg/plugins/utils/update_upstream_spec_test.go
@@ -57,6 +57,7 @@ var _ = Describe("UpdateUpstream", func() {
                desiredFailover := &gloov1.Failover{PrioritizedLocalities: []*gloov1.Failover_PrioritizedLocality{{}}}
                desiredUseHttp2 := &wrappers.BoolValue{Value: true}
                desiredHttpProxyHostname := &wrappers.StringValue{Value: "desiredHostname"}
+               desiredHttpProxyHeaders := []*gloov1.HeaderValue{{Key: "k", Value: "v"}}
                desired := &gloov1.Upstream{
                        SslConfig:          desiredSslConfig,
                        CircuitBreakers:    desiredCircuitBreaker,
@@ -67,6 +68,7 @@ var _ = Describe("UpdateUpstream", func() {
                        Failover:           desiredFailover,
                        UseHttp2:           desiredUseHttp2,
                        HttpProxyHostname:  desiredHttpProxyHostname,
+                       HttpConnectHeaders: desiredHttpProxyHeaders,
                }
                original := &gloov1.Upstream{
                        SslConfig:          &gloov1.UpstreamSslConfig{Sni: "testsni"},
@@ -78,6 +80,7 @@ var _ = Describe("UpdateUpstream", func() {
                        Failover:           &gloov1.Failover{PrioritizedLocalities: []*gloov1.Failover_PrioritizedLocality{{}, {}}},
                        UseHttp2:           &wrappers.BoolValue{Value: false},
                        HttpProxyHostname:  &wrappers.StringValue{Value: "originalHostname"},
+                       HttpConnectHeaders: desiredHttpProxyHeaders,
                }
 
                utils.UpdateUpstream(original, desired)
@@ -90,6 +93,7 @@ var _ = Describe("UpdateUpstream", func() {
                Expect(desired.Failover).To(Equal(desiredFailover))
                Expect(desired.UseHttp2).To(Equal(desiredUseHttp2))
                Expect(desired.HttpProxyHostname).To(Equal(desiredHttpProxyHostname))
+               Expect(desired.HttpConnectHeaders).To(Equal(desiredHttpProxyHeaders))
        })
 
        It("will fail if the upstream proto has a new top level field", func() {
@@ -97,7 +101,7 @@ var _ = Describe("UpdateUpstream", func() {
                // This should happen very rarely, and should be used as an indication that the `UpdateUpstream` function
                // most likely needs to change.
                Expect(reflect.TypeOf(gloov1.Upstream{}).NumField()).To(
-                       Equal(23),
+                       Equal(24),
                        "wrong number of fields found",
                )
        })
```

pwoelfle commented:
@kdorosh Patch applied 👍🏻

kdorosh commented:
/test flake? https://storage.googleapis.com/solo-public-build-logs/logs.html?buildid=a6ac8ff7-6909-4b21-801e-f79d2a7b31f0

kdorosh commented:
```
Step #10 - "test": [2022-08-19 14:13:05.446][143290][debug][http] [external/envoy/source/common/http/filter_manager.cc:895] [C2][S5890064527930615980] Sending local reply with details cluster_not_found
```

...

```
Step #10 - "test": [2022-08-19 14:13:05.463][143244][debug][config] [external/envoy/source/common/config/grpc_subscription_impl.cc:113] gRPC update for type.googleapis.com/envoy.config.route.v3.RouteConfiguration failed
Step #10 - "test": SSSSSSSSSSSSSSSSSSSSSSSSS
Step #10 - "test": JUnit report was created: /workspace/gloo/test/e2e/junit.xml
Step #10 - "test": 
Step #10 - "test": 
Step #10 - "test": Summarizing 1 Failure:
Step #10 - "test": 
Step #10 - "test": [Fail] tunneling with Proxy Authorization when using invalid credentials [It] should not proxy 
Step #10 - "test": /workspace/gloo/test/e2e/http_tunneling_test.go:298
Step #10 - "test": 
Step #10 - "test": Ran 134 of 159 Specs in 486.659 seconds
Step #10 - "test": FAIL! -- 133 Passed | 1 Failed | 0 Pending | 25 Skipped
Step #10 - "test": --- FAIL: TestE2e (489.35s)
Step #10 - "test": FAIL
Step #10 - "test": 
Step #10 - "test": Ginkgo ran 127 suites in 28m38.656825012s
Step #10 - "test": Test Suite Failed
Step #10 - "test": make: *** [Makefile:170: run-tests] Error 1
Finished Step #10 - "test"
ERROR
```

test failure looks real and related to this PR (was introduced here). I was able to reproduce this locally every once in a while; appears to be a flaky test that fails more reliably in ci. @pwoelfle can you try to deflake?

pwoelfle commented:
@kdorosh I've fixed the flaky tests.

The reason was that in some cases the Envoy didn't know the cluster. Sometimes, there seems to be some delay while applying the configurations. Then, the test request receives also a HTTP 503 Service Unavailable due to the unknown cluster, but the response body is empty. 

Here you can see the Envoy log of such an "unknown cluster" test request:
```console
[2022-08-22 05:54:58.001][17][debug][conn_handler] [external/envoy/source/server/active_tcp_listener.cc:141] [C2] new connection from 172.17.0.1:61340
[2022-08-22 05:54:58.002][17][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:304] [C2] new stream
[2022-08-22 05:54:58.002][17][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:909] [C2][S12532456944200730562] request headers complete (end_stream=false):
':authority', 'localhost:11086'
':path', '/test'
':method', 'POST'
'user-agent', 'Go-http-client/1.1'
'content-length', '25'
'accept-encoding', 'gzip'

[2022-08-22 05:54:58.002][17][debug][connection] [external/envoy/source/common/network/connection_impl.h:89] [C2] current connecting state: false
[2022-08-22 05:54:58.002][17][debug][router] [external/envoy/source/common/router/router.cc:454] [C2][S12532456944200730562] unknown cluster 'local-1_default'
[2022-08-22 05:54:58.002][17][debug][http] [external/envoy/source/common/http/filter_manager.cc:895] [C2][S12532456944200730562] Sending local reply with details cluster_not_found
[2022-08-22 05:54:58.002][17][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:1525] [C2][S12532456944200730562] encoding headers via codec (end_stream=true):
':status', '503'
'date', 'Mon, 22 Aug 2022 05:54:57 GMT'
'server', 'envoy'
'connection', 'close'

```

kdorosh commented:
@pwoelfle our code formatting check here failed: https://github.com/solo-io/gloo/runs/7955826426?check_suite_focus=true

```
index c3d17933d..0eb409037 100644
--- a/test/e2e/http_tunneling_test.go
+++ b/test/e2e/http_tunneling_test.go
@@ -8,7 +8,6 @@ import (
 	"encoding/base64"
 	"errors"
 	"fmt"
-	"github.com/onsi/gomega/types"
 	"io"
 	"net"
 	"net/http"
@@ -18,6 +17,8 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/onsi/gomega/types"
+
 	"github.com/solo-io/gloo/test/v1helpers"
 
 	"github.com/golang/protobuf/ptypes/wrappers"
```

can you run `goimports -w ./` and commit?

pwoelfle commented:
@kdorosh Correct formatting applied.

kdorosh commented:
/test

kdorosh commented:
/skip-ci error on forks, will fix in follow up. unrelated to this feature

kdorosh commented:
PR to sort out other test errors from forks https://github.com/solo-io/gloo/pull/7009, just fyi

pwoelfle commented:
@kdorosh Thanks for your support!
